src/common/entity/compute_state_display.ts

```diff
@@ -29,37 +29,69 @@ export const computeStateDisplay = (
   const domain = computeStateDomain(stateObj);
 
   if (domain === "input_datetime") {
-    let date: Date;
-    if (!stateObj.attributes.has_time) {
+    if (state) {
+      // If trying to display an explicit state, need to parse the explict state to `Date` then format.
+      // Attributes aren't available, we have to use `state`.
+      try {
+        if (!stateObj.attributes.has_time) {
+          // Only has date.
+          const dateObj = new Date(state);
+          // When only date is passed to `Date` constructor, it uses UTC regardless of frontend's timezone,
+          // so we need to add the frontend's timezone offset.
+          // Other usages of `Date` constructor have both date and time in the string, frontend's timezone is used w/o problem.
+          const timezoneOffset = new Date().getTimezoneOffset() * 60 * 1000;
+          const adjustedDateObj = new Date(dateObj.getTime() + timezoneOffset);
+          return formatDate(adjustedDateObj, locale);
+        }
+        if (!stateObj.attributes.has_date) {
+          // Only has time.
+          const now = new Date();
+          const dateTiemString = now.toDateString() + " " + state;
+          const dateObj = new Date(dateTiemString);
+          return formatTime(dateObj, locale);
+        }
+        // Has both date and time.
+        const dateObj = new Date(state);
+        return formatDateTime(dateObj, locale);
+      } catch {
+        // If `Date` constructor throws error, meaning the explict state isn't a valid date/time string,
+        // just return it.
+        return state;
+      }
+    } else {
+      // If not trying to display an explicit state, create `Date` object from `stateObj`'s attributes then format.
+      let date: Date;
+      if (!stateObj.attributes.has_time) {
+        date = new Date(
+          stateObj.attributes.year,
+          stateObj.attributes.month - 1,
+          stateObj.attributes.day
+        );
+        return formatDate(date, locale);
+      }
+      if (!stateObj.attributes.has_date) {
+        const now = new Date();
+        date = new Date(
```

Comment:
Instead of creating a new date, we can just set the hour and minute of now?

Reply:
This is done.